numba/misc/dummyarray.py

```diff
@@ -164,7 +165,47 @@ def __init__(self, dims, itemsize):
         self.itemsize = itemsize
         self.size = functools.reduce(operator.mul, self.shape, 1)
         self.extent = self._compute_extent()
-        self.flags = self._compute_layout()
+        if config.NPY_RELAXED_STRIDES_CHECKING:
+            self.flags = self._relaxed_compute_layout()
+        else:
+            self.flags = self._compute_layout()
+
+    def _relaxed_compute_layout(self):
+        # The logic here is based on that in _UpdateContiguousFlags from
+        # numpy/core/src/multiarray/flagsobject.c in NumPy v1.19.1 (commit
+        # 13661ac70).
+
+        # Records have no dims, and we can treat them as contiguous
+        if not self.dims:
+            return {'C_CONTIGUOUS': True, 'F_CONTIGUOUS': True}
+
+        # If this is a broadcast array then it is not contiguous
+        if any([dim.stride == 0 for dim in self.dims]):
+            return {'C_CONTIGUOUS': False, 'F_CONTIGUOUS': False}
```

Comment:
Are these branches tested?

Reply:
Yes, the test is `numba.cuda.tests.cudadrv.test_cuda.ndarray.TestCudaNDArray.test_devicearray_broadcast_host_copy`. The test seems quite sensitive to mistakes in contiguity computation - it took a while to arrive at the correct logic, which is partly why I wanted to have a flag to go back to strict strides checking - this function is correct to the best of my ability and knowledge now, but there may well be blind spots in the test suite and my knowledge.